crates/holochain/src/core/workflow/countersigning_workflow/tests.rs

```diff
@@ -410,6 +410,143 @@ async fn receive_valid_and_invalid_signatures_and_complete() {
+
+            net.expect_publish_countersign()
+                .times(2)
+                .returning(|_, _, _| Ok(()));
+
+            net
+        }
+    });
+
+    let bob = test_harness.new_remote_agent().await;
+
+    let request = test_preflight_request(&test_harness, Duration::from_secs(60), &bob);
+    let my_acceptance = test_harness
+        .accept_countersigning_request(request.clone())
+        .await
+        .unwrap();
+
+    test_harness
+        .respond_to_countersigning_workflow_signal()
+        .await;
+
+    let bob_acceptance = bob
+        .accept_preflight_request(request.clone(), test_harness.keystore.clone())
+        .await;
+
+    let (session_data, entry, entry_hash) =
+        test_harness.build_session_data(request.clone(), vec![my_acceptance, bob_acceptance]);
+
+    // Receive the signatures for this session.
+    let signatures = vec![
+        bob.produce_signature(&session_data, &entry_hash, test_harness.keystore.clone())
+            .await,
+        test_harness
+            .commit_countersigning_entry(&session_data, entry.clone(), entry_hash.clone())
+            .await,
+    ];
+    countersigning_success(
+        test_harness.test_space.space.clone(),
+        test_harness.author.clone(),
+        signatures.clone(),
+        test_harness.countersigning_tx.clone(),
+    )
+    .await;
+
+    test_harness
+        .respond_to_countersigning_workflow_signal()
+        .await;
+
+    // The session gets completed by the first signature bundle we received
+    test_harness.expect_success_signal().await;
+    test_harness.expect_publish_and_integrate();
+    test_harness.expect_no_pending_signals();
+    test_harness.expect_empty_workspace();
+    test_harness.expect_scheduling_complete();
+
+    // Now we can start a new session
+    let new_request = test_preflight_request(&test_harness, Duration::from_secs(60), &bob);
+    let new_my_acceptance = test_harness
+        .accept_countersigning_request(new_request.clone())
+        .await
+        .unwrap();
+
+    test_harness
+        .respond_to_countersigning_workflow_signal()
+        .await;
+
+    let new_bob_acceptance = bob
+        .accept_preflight_request(new_request.clone(), test_harness.keystore.clone())
+        .await;
+
+    let (session_data, entry, entry_hash) = test_harness.build_session_data(
+        new_request.clone(),
+        vec![new_my_acceptance, new_bob_acceptance],
+    );
+
+    test_harness.expect_session_accepted();
+
+    // Receive the previous signatures again, they should be ignored.
+    countersigning_success(
+        test_harness.test_space.space.clone(),
+        test_harness.author.clone(),
+        signatures.clone(),
+        test_harness.countersigning_tx.clone(),
+    )
+    .await;
+
+    test_harness
+        .respond_to_countersigning_workflow_signal()
+        .await;
+
+    // Is this ideal? We're accepting the signatures
+    test_harness.expect_session_in_signatures_collected();
```

Comment:
Do we assert that the countersigning workflow is still in progress? Is that what this call does? Are we sure that it isn't already in a success state at this point? Just trying to get my head around it.

Reply:
The workflow operates a state machine. A state change has happened here and we should be "waiting for more signatures". So yes this is reading the current state of the workflow. If we'd completed the session somehow here then there would be no state, rather than this state. Also, the following steps would fail.

What this is really here for is to say that we didn't stay in "Accepted" and ignore this input. We do accept bad signature bundles and only find out when we try to use them that they aren't good.